ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -31,215 +81,264 @@ where
+            }
+            let p_extrapolated = lagrange_table.eval(&uv.borrow().0);
+            let q_extrapolated = lagrange_table.eval(&uv.borrow().1);
+
+            for (i, (x, y)) in
+                zip(p_extrapolated.into_iter(), q_extrapolated.into_iter()).enumerate()
+            {
+                proof.g[R::USIZE + i] += x * y;
+            }
+        }
+        proof
+    }
 }
 
 #[derive(Debug)]
-pub struct ProofGenerator<F: PrimeField> {
-    u: Vec<F>,
-    v: Vec<F>,
+pub struct UVStore<F, R>
+where
+    F: PrimeField,
+    R: ArrayLength,
+{
+    uv: Vec<UVPolynomial<F, R>>,
 }
 
-pub type TwoNMinusOne<N> = Diff<Sum<N, N>, U1>;
-pub type TwoNPlusOne<N> = Sum<Sum<N, N>, U1>;
-
-///
-/// Distributed Zero Knowledge Proofs algorithm drawn from
-/// `https://eprint.iacr.org/2023/909.pdf`
-///
-#[allow(non_camel_case_types, clippy::many_single_char_names)]
-impl<F> ProofGenerator<F>
+impl<F, R> UVStore<F, R>
 where
     F: PrimeField,
+    R: ArrayLength,
 {
-    pub fn new(u: Vec<F>, v: Vec<F>) -> Self {
-        debug_assert_eq!(u.len(), v.len(), "u and v must be of equal length");
-        Self { u, v }
+    pub fn len(&self) -> usize {
+        self.uv.len() * R::USIZE
     }
 
-    pub fn compute_proof<λ, I, J>(
-        u: I,
-        v: J,
-        lagrange_table: &LagrangeTable<F, λ, <λ as Sub<U1>>::Output>,
-    ) -> ZeroKnowledgeProof<F, TwoNMinusOne<λ>>
-    where
-        λ: ArrayLength + Add + Sub<U1>,
-        <λ as Add>::Output: Sub<U1>,
-        <<λ as Add>::Output as Sub<U1>>::Output: ArrayLength,
-        <λ as Sub<U1>>::Output: ArrayLength,
-        I: IntoIterator<Item = F>,
-        J: IntoIterator<Item = F>,
-        I::IntoIter: ExactSizeIterator,
-        J::IntoIter: ExactSizeIterator,
-    {
-        let mut u = u.into_iter();
-        let mut v = v.into_iter();
+    pub fn is_empty(&self) -> bool {
+        self.uv.is_empty()
+    }
 
-        debug_assert_eq!(u.len() % λ::USIZE, 0); // We should pad with zeroes eventually
+    pub fn iter(&self) -> impl Iterator<Item = &UVPolynomial<F, R>> + Clone {
+        self.uv.iter()
+    }
 
-        let s = u.len() / λ::USIZE;
+    pub fn first(&self) -> &UVPolynomial<F, R> {
+        &self.uv[0]
+    }
 
-        assert!(
-            s > 1,
-            "When the output is this small, you should call `compute_final_proof`"
+    /// This function allows to set masks to protect sensitive information.
+    /// It is used when the last proof of a batch is generated.
+    ///
+    /// ## Panics
+    /// Panics when `self` contains more than one polynomial tuple
+    /// or when masks cannot be set safely.
+    /// The latter is the case when there are too many points for a polynomial of degree `R minus 1`.
+    /// This function likely panics when used with small fields like `Fp31`
+    /// Use `FP61BitPrime` instead when using this function in test cases.
+    pub fn set_masks(&mut self, p_0: F, q_0: F) {
```

Comment:
is there a reason for not using `Result`? That way you could easily test the behaviour and panic in a single place